snowfall/training/asg_graph.py

```diff
@@ -0,0 +1,164 @@
+    if 0 in ans:
+        ans.remove(0)
+    ans.sort()
+    return ans
+
+
+def create_bigram_phone_lm(phones: List[int]) -> k2.Fsa:
+    '''Create a bigram phone LM.
+    The resulting FSA (P) has a start-state and a state for
+    each phone 1, 2, ....; and each of the above-mentioned states
+    has a transition to the state for each phone and also to the final-state.
+
+    Caution:
+      blank is not a phone.
+
+    Args:
+      A list of phone IDs.
+
+    Returns:
+      An FSA representing the bigram phone LM.
+    '''
+    assert 0 not in phones
+    final_state = len(phones) + 1
+    rules = ''
+    for i in range(1, final_state):
+        rules += f'0 {i} {phones[i-1]} 0.0\n'
+
+    for i in range(1, final_state):
+        for j in range(1, final_state):
+            rules += f'{i} {j} {phones[j-1]} 0.0\n'
+        rules += f'{i} {final_state} -1 0.0\n'
+    rules += f'{final_state}'
+    return k2.Fsa.from_str(rules)
+
+
+class AsgTrainingGraphCompiler(object):
+
+    def __init__(self,
+                 L_inv: k2.Fsa,
+                 phones: k2.SymbolTable,
+                 words: k2.SymbolTable,
+                 oov: str = '<UNK>'):
+        '''
+        Args:
+          L_inv:
+            Its labels are words, while its aux_labels are phones.
+        phones:
+          The phone symbol table.
+        words:
+          The word symbol table.
+        oov:
+          Out of vocabulary word.
+        '''
+
+        if L_inv.properties & k2.fsa_properties.ARC_SORTED != 0:
+            L_inv = k2.arc_sort(L_inv)
+
+        assert L_inv.requires_grad is False
+
+        assert oov in words
+
+        self.L_inv = L_inv
+        self.phones = phones
+        self.words = words
+        self.oov = oov
+
+        phone_symbols = get_phone_symbols(phones)
+        phone_symbols_with_blank = [0] + phone_symbols
+
+        ctc_topo = build_ctc_topo(phone_symbols_with_blank)
+        assert ctc_topo.requires_grad is False
+
+        self.ctc_topo = ctc_topo
+
+    def compile(self, texts: Iterable[str],
+                P: k2.Fsa) -> Tuple[k2.Fsa, k2.Fsa]:
+        '''Create numerator and denominator graphs from transcripts
+        and the bigram phone LM.
+
+        Args:
+          texts:
+            A list of transcripts. Within a transcript, words are
+            separated by spaces.
+          P:
+            The bigram phone LM created by :func:`create_bigram_phone_lm`.
+        Returns:
+          A tuple (num_graph, den_graph), where
+
+            - `num_graph` is the numerator graph. It is an FsaVec with
+              shape `(len(texts), None, None)`.
+
+            - `den_graph` is the denominator grap. It is an FsaVec with the same
+              shape of the `num_graph`.
+        '''
+        assert P.is_cpu()
+
+        den = k2.intersect(self.ctc_topo, P).invert_()
+        den = k2.connect(den)
+
+        decoding_graphs = k2.create_fsa_vec(
+            [self.compile_one_and_cache(text) for text in texts])
+
```

Comment:
Where are you composing the num graph with P?  We can't train the probs on the den graph only, they'll be backwards.
In fact, my plan, in order to get LM-like probs, was to block the derivative from the den graph (using .detach() or something) and only use the derivative from the num graph.
Note: this is  incompatible with caching, we need to compose the num graph with P after caching.
Also I'm wondering where the topology comes in, e.g. the CTC topology or whatever topology we're using that allows for self-loops?

Reply:
.. as far as I can tell ytou are including the CTC topo in the den graph but not the num graphs; same with P.

Reply:
> Where are you composing the num graph with P? 

https://github.com/k2-fsa/snowfall/blob/33cf6fedb071e31fdabbe11dbc39d9adfe8a8ada/snowfall/training/asg_graph.py#L139

Is it correct?

Reply:
also please call the num graphs `num_graphs` not `decoding_graphs`.. a decoding graph would normally have some kind of LM, if anything this is an alignment graph but in thiws context let's call it a num_graphl

Reply:
Ah, yes, I suppose that could be correct, assuming everything is the right way round.
You might want to verify that the probs in P are being trained and look OK, e.g. more common phones have higher prob than less common ones.

Reply:
> please call the num graphs num_graphs not decoding_graphs

thanks, will change it.